### PR TITLE
📦 NEW: Improve data-gen, train new policy

### PR DESCRIPTION
**Pull request by ianagbip1oti**
```
data/princhess-20260628-2219-401m.data.shuffled
```

```
--------------------------------------------------
Results of princhess vs princhess-main (40+0.4, 1t, 128MB, UHO_Lichess_4852_v1.epd):
Elo: 9.37 +/- 5.58, nElo: 17.91 +/- 10.66
LOS: 99.95 %, DrawRatio: 52.18 %, PairsRatio: 1.24
Games: 4078, Wins: 916, Losses: 806, Draws: 2356, Points: 2094.0 (51.35 %)
Ptnml(0-2): [22, 413, 1064, 513, 27], WL/DD Ratio: 0.49
LLR: 2.90 (100.2%) (-2.25, 2.89) [0.00, 3.00]
--------------------------------------------------

--------------------------------------------------
Results of princhess vs princhess-main (40+0.4, 1t, 128MB, DFRC_4852_v1.epd):
Elo: 6.22 +/- 5.53, nElo: 10.80 +/- 9.61
LOS: 98.63 %, DrawRatio: 47.63 %, PairsRatio: 1.12
Games: 5026, Wins: 1160, Losses: 1070, Draws: 2796, Points: 2558.0 (50.90 %)
Ptnml(0-2): [50, 572, 1197, 626, 68], WL/DD Ratio: 0.50
LLR: 2.90 (100.5%) (-2.25, 2.89) [-3.00, 0.00]
--------------------------------------------------

--------------------------------------------------
Results of princhess vs princhess-main (8+0.08, 1t, 128MB, UHO_Lichess_4852_v1.epd):
Elo: 2.22 +/- 3.84, nElo: 3.80 +/- 6.59
LOS: 87.08 %, DrawRatio: 47.01 %, PairsRatio: 1.07
Games: 10662, Wins: 2548, Losses: 2480, Draws: 5634, Points: 5365.0 (50.32 %)
Ptnml(0-2): [146, 1219, 2506, 1341, 119], WL/DD Ratio: 0.63
LLR: 2.89 (100.1%) (-2.25, 2.89) [-3.00, 0.00]
--------------------------------------------------

```

## Summary by CodeRabbit

* **New Features**
  * Added new training and benchmarking commands, including support for alternate test variants and engine comparisons.
  * Introduced a live terminal experience for training with progress, metrics, and checkpoint status.
  * Added scripts to start, resume, and generate tuning runs more easily.

* **Bug Fixes**
  * Updated Docker and runtime handling to avoid modifying mounted config files and to reduce build context size.
  * Adjusted defaults for engine settings and matchmaking timing for faster responses.

* **Documentation**
  * Clarified setup, build, and testing guidance.

### CODE REVIEW

File: bin/Dockerfile.lichess [hunk 1]
```diff
@@ -21,9 +21,8 @@ RUN  git clone https://github.com/ShailChoksi/lichess-bot.git /src \
  && git checkout 19207a5fbcfb2462da0eeb1004af897e9270afb7 \
  && pip install -r requirements.txt
 
-COPY bin/config.yml /src/config.yml
 COPY bin/human.bin /books/human.bin
 COPY --from=build /src/target/release/princhess /engines/princhess
 
-CMD /run-bot.sh
+CMD ["/run-bot.sh"]
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply bin/Dockerfile.lichess:27]: This is a deliberate change, the config is generated dynamically, so is mounted when it is run via make tune

File: crates/princhess-train/src/args.rs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        }
+    }
+
+    /// Expect a positional argument or panic
+    ///
+    /// # Panics
+    /// Panics if no argument is available
+    pub fn expect(&mut self, name: &str) -> String {
+        self.next()
+            .unwrap_or_else(|| panic!("Missing required argument: {name}"))
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/args.rs:60]: Claude says: 

This is intentional. The Args parser follows a two-phase pattern: all flag() calls happen first (each consuming its own key and value from the arg list), leaving only positional arguments. expect() then consumes
  those positionals in order.

  For these developer-only CLI tools with a handful of known flags, the simplicity is worth the tradeoff. A misspelled flag like --phse value would be surfaced as the positional — confusing, but the failure is loud:
  either the flag's effect is absent (defaulted) or expect() picks up the wrong value and a downstream error occurs. There's no silent success.

  Adding flag-prefix detection (if value.starts_with('-')) to expect() would be a reasonable hardening step but adds complexity we don't currently need. If the CLI surface grows or becomes more user-facing, that's
  the right time to revisit.

File: crates/princhess-train/src/bin/train-policy.rs [hunk 3]
```diff
@@ -707,52 +937,53 @@ fn render_accuracy_chart(frame: &mut Frame, area: ratatui::layout::Rect, stats:
...
 
-            let batch_metrics = gradients_batch(network, &mut gradients, &mut count, batch, phase);
+            let batch_metrics =
+                gradients_batch(network, &mut gradients, batch, &mut thread_buffers, config, phase);
 
-            gradients.scale_by_counts(&count);
+            stats.record_batch(batch_metrics);
 
             optimizer.step();
             network.adamw(&gradients, momentum, velocity, optimizer);
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/bin/train-policy.rs]: BATCH_SIZE is 32,768. Phase filtering splits chess positions into mg/eg, and real training data covers both phases throughout. The probability of all 32,768 positions in a batch failing a phase check is effectively
   zero — even an extreme skew of 99% against the target phase would still leave ~328 hits per batch.

  The gradients_batch function already handles processed_count == 0 safely (skips the division at line 1054), so the only theoretical harm would be an adamw update with zero gradients, which just advances the step
  counter and applies weight decay — a cosmetic waste on a scenario that never occurs in practice. Not worth the complexity.

File: crates/princhess-train/src/bin/train-policy.rs [hunk 4]
```diff
@@ -855,30 +1082,40 @@ fn update_gradient(
...
-            + (actual_val - expected_secondary_val) * soft_target_weight;
-        network.backprop(&features, gradients, move_idx, combined_error);
-        count.increment(move_idx);
+        network.backprop(&features, gradients, move_idxes[idx], error);
     }
 
-    if argmax(&expected_primary) == argmax(&actual_policy) {
+    let baseline = (moves.len() as f32).ln();
+    metrics.loss += position_loss;
+    metrics.baseline_loss += baseline;
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/bin/train-policy.rs:1134]: info gain is used comparatively, not as an absolute calibrated measure. The 10% constant offset from the soft target weight doesn't affect relative comparisons between runs or
  checkpoints, and fixing it would change reported values without improving any decision we make from them.

File: crates/princhess-train/src/neural/lr_scheduler.rs [hunk 5]
```diff
@@ -1,3 +1,34 @@
...
+}
+
+impl LRScheduler for LinearWarmupDecayLRScheduler {
+    fn get_lr(&self, step: u32) -> f32 {
+        if step < self.warmup_steps {
+            self.initial_lr * step as f32 / self.warmup_steps as f32
+        } else {
+            let decay_steps = self.total_steps - self.warmup_steps;
+            let elapsed = step - self.warmup_steps;
+            self.initial_lr * (1.0 - elapsed as f32 / decay_steps as f32).max(0.0)
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/neural/lr_scheduler.rs:27]: internal developer API with fixed constants, no user input reaches these constructors, and the invalid configurations are self-evidently wrong. An assert
  would only fire if someone misuses the constructor in a new callsite, at which point the panic message from the division itself is equally informative.

File: Makefile [hunk 6]
```diff
@@ -6,36 +6,61 @@ help:
...
+
+TRAIN_FEATURES = $(if $(NODEFAULTFEATURES),--no-default-features,--features princhess/default)
+
+.PHONY: train
+train:
+	@if [ -z "$(FORCE)" ] && [ -z "$(NODEFAULTFEATURES)" ] && ! git diff --quiet main -- src/nets/*.bin 2>/dev/null; then \
+		echo "WARNING: Net files differ from main branch!"; \
+		echo "Use 'make train NODEFAULTFEATURES=1' to build without default features (e.g. after changing net struct size)."; \
+		echo "Use 'make train FORCE=1' to build with default features anyway."; \
+		exit 1; \
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply Makefile:57]: if you can't check against main, you also can't verify the nets are correct, so blocking is the right behavior. "Main not found" is a suspicious situation (fresh clone, detached HEAD,
  wrong remote setup) and silently continuing could let a mismatched net slip through.

File: scripts/generate-tuning-config.sh [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+validate_option() {
+    local option_name=$1
+    if ! echo "$UCI_OUTPUT" | grep -q "option name $option_name "; then
+        err "Error: Unknown UCI option '$option_name'"
+        err ""
+        err "Available options:"
+        echo "$UCI_OUTPUT" | grep "option name" | sed 's/option name \([^ ]*\).*/  \1/' >&2
+        exit 1
+    fi
+}
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply scripts/generate-tuning-config.sh:98]: If a non-spin option (e.g. check, string) is passed, calculate_range tries bash arithmetic on the default value (e.g. "true"), which fails and exits under set -e. The script already can't produce invalid JSON
  output; it just crashes with a confusing error instead of a clean one. This is a developer-facing script where the caller knows they're tuning spin parameters — worth skipping.

File: crates/princhess-train/src/bin/data-gen.rs [hunk 8]
```diff
@@ -1098,28 +1285,31 @@ fn render_tui(frame: &mut Frame, view: &StatsView) {
...
+        &bucket_labels(&EVAL_DISTRIBUTION_THRESHOLDS),
     );
     render_histogram(
         frame,
         histogram_row[3],
-        "Eval-Result Agreement",
+        &format!(
+            "Eval-Result Agreement (μ={:.2})",
+            view.avg_eval_result_disagreement()
+        ),
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/bin/data-gen.rs:1306]: Fixed in https://github.com/princesslana/princhess/pull/477/commits/272663b60400d0f9df7f477405ffae569cce65b8

File: crates/princhess-train/src/bin/data-gen.rs [hunk 9]
```diff
@@ -1302,24 +1512,50 @@ fn run_tui(stats: &Stats, stop_signal: Arc<AtomicBool>) -> io::Result<()> {
...
+        threads > 0,
+        "Thread count must be at least 1, got {threads}"
+    );
+    assert!(
+        max_positions <= MAX_POSITIONS_PER_FILE * u64::from(threads),
+        "Requested {max_positions} positions but maximum is {} ({}M positions/file * {} threads)",
+        MAX_POSITIONS_PER_FILE * u64::from(threads),
+        MAX_POSITIONS_PER_FILE / 1_000_000,
+        threads
+    );
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/bin/data-gen.rs:1531]: Fixed in https://github.com/princesslana/princhess/pull/477/commits/272663b60400d0f9df7f477405ffae569cce65b8

File: crates/princhess-train/src/bin/data-gen.rs [hunk 10]
```diff
@@ -1302,24 +1512,50 @@ fn run_tui(stats: &Stats, stop_signal: Arc<AtomicBool>) -> io::Result<()> {
...
                 eprintln!("TUI failed: {e}");
                 tui_stop.store(true, Ordering::Relaxed);
             }
         });
 
-        for t in 0..THREADS {
-            thread::sleep(Duration::from_millis(10 * t));
+        stats
+            .active_threads
+            .store(threads as usize, Ordering::Relaxed);
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/bin/data-gen.rs]: Fixed in https://github.com/princesslana/princhess/pull/477/commits/272663b60400d0f9df7f477405ffae569cce65b8

File: scripts/bench.sh [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+# Compare NPS between two engine builds by running bench multiple times.
+
+set -e
+
+SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+PROJECT_ROOT="$(dirname "$SCRIPT_DIR")"
+
+ENGINE1="${1:-princhess}"
+ENGINE2="${2:-princhess-main}"
+RUNS="${3:-10}"
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply scripts/bench.sh:12]: Fixed in https://github.com/princesslana/princhess/pull/477/commits/272663b60400d0f9df7f477405ffae569cce65b8
